db/bloc_record/lib/bloc_record/selection.rb

```diff
@@ -1,22 +1,138 @@
 require 'sqlite3'
 
 module Selection
-  def find(id)
+  def find_one(id)
+  	validate_integer(id)
     row = connection.get_first_row <<-SQL
 	    SELECT #{columns.join ","} FROM #{table}
 	    WHERE id = #{id};
-    SQL
+	  SQL
+    init_object_from_row(row)
+  end
 
-    data = Hash[columns.zip(row)]
-    new(data)
+  def find(*ids)
+  	ids.each do |id|
+  		validate_integer(id)
+  	end
+  	if id.length == 1
+  		find_one(ids.first)
+  	else
+  		rows = connection.execute <<-SQL 
+  			SELECT #{columns.join ","} FROM #{table}
+  			WHERE id IN (#{ids.join ","});
+  		SQL
+  	end
+  	rows_to_array(rows)
   end
-end
 
+  def find_by(attribute, value)
+  	row = connection.get_first_row <<-SQL 
+  		SELECT #{columns.join(",")} from #{table}
+  		WHERE #{attribute} = #{BlocRecord::Utility.sql_strings(value)};
+  	SQL
+
+  	init_object_from_row(row)
+  end
+
+  def take_one
+  	row = connection.get_first_row <<-SQL
+  		SELECT #{columns.join ","} FROM #{table}
+  		ORDER BY random()
+  		LIMIT 1;
+  	SQL
+  end
+
+  def first
+  	row = connection.get_first_row <<-SQL 
+  		SELECT #{columns.join ","} FROM #{table}
+  		ORDER BY id 
+  		ASC LIMIT 1;
+  	SQL
+
+  	init_object_from_row(row)
+  end
+
+  def last
+  	row = connection.get_first_row <<-SQL 
+  		SELECT #{columns.join ","} FROM #{table}
+  		ORDER BY id 
+  		desc LIMIT 1;
+  	SQL
+
+  	init_object_from_row(row)
+  end
+
+  def take(num=1)
+   validate_integer(num)
+   if num > 1
+     rows = connection.execute <<-SQL
+       SELECT #{columns.join ","} FROM #{table}
+       ORDER BY random()
+       LIMIT #{num};
+     SQL
+
+     rows_to_array(rows)
+   else
+     take_one
+   end
+	end
+
+	def all
+		rows = connection.execute <<-SQL 
+			SELECT #{columns.join ","} FROM #{table};
+		SQL
 
-def find_by(attribute, value)
-	results = []
-	connection.execute("SELECT * FROM #{table} WHERE #{attribute}=#{value};") do |row|
-		row << results
+		rows_to_array(rows)
 	end
-	return results
-end
\ No newline at end of file
+
+  def find_each(options={}, &block)
+    rows = connection.execute <<-SQL 
+      SELECT #{columns.join ","} FROM #{table}
+      LIMIT #{options[:batch_size]}
+    SQL
+
+    rows_array = rows_to_array(rows)
+
+    if block_given?
+      rows_array.each do |row|
+        yield(row)
+      end
+    end
+    rows_array
+  end  
+
+  def find_in_batches(options={}, &block)
+    batch = find_each(start: options[:start], batch_size: options[:batch_size])
+    object_array = []
+    batch.each do |row|
+      object_array << init_object_from_row(row)
+    end
+
+    if block_given?
+      yield(object_array, batch)
+    end 
+
+    object_array
+  end 
+
+	private
+
+	def validate_integer(int)
+		raise "ID cannot be negative" if int < 0
```

Comment:
- [ ] I'm not sure you would want to raise an exception here. It's probably OK, but you'd need to handle every possible instance of it being thrown gracefully otherwise it will bring your application to a screeching halt.